pkg/frontend/transport/seeker_body.go

```diff
@@ -0,0 +1,54 @@
+// SPDX-License-Identifier: AGPL-3.0-only
+
+package transport
+
+import (
+	"bytes"
+	"fmt"
+	"io"
+	"net/http"
+)
+
+// ParseSeekerBodyForm will ensure that request body is a Seeker, parse the form and seek the body back to the beginning.
+func ParseSeekerBodyForm(r *http.Request) error {
+	return withSeekerBody(r, func(r *http.Request) error {
+		return r.ParseForm()
+	})
+}
+
+// withSeekerBody will ensure that request body is a Seeker, call the provided function and seek the body back to the beginning.
+func withSeekerBody(r *http.Request, f func(r *http.Request) error) (err error) {
```

Comment:
It is restored, the `defer` seeks the body to the beginning.

Reply:
What I mean is that this function (possilby) replaces `r.Body` as a side effect, then calls `f` and then (!) resets body back to beginning. I think it does too many things, and they are not in the logical order.

I would separate these three concerns:
- we already replace `r.Body` in Handler's `ServeHTTP` method. No need to do it again here.
- `withSeekerBody` should just 1) check if body is seeker, and 2) reset it (ie. it becomes `resetSeekerBody`). If body is not seeker, return error.
- code using the body can stay as it was before, and not called indirectly via function

Reply:
> then calls `f` and then (!) resets body back to beginning.

Just a note: I realize this makes sense for usage from middlewares, where we want to make sure that request can be used by next step. Comments in the code explaining why we want this would be appreciated, as it seems weird.